analyzers/src/SonarAnalyzer.Common/Rules/Utilities/UtilityAnalyzerBase.cs

```diff
@@ -92,34 +93,39 @@ public abstract class UtilityAnalyzerBase<TMessage> : UtilityAnalyzerBase
                         return;
                     }
 
-                    // The results of Metrics and CopyPasteToken analyzers are not needed for Test projects yet the plugin side expects the protobuf files, so we create empty ones.
                     if (IsTestProject && SkipAnalysisForTestProject)
                     {
-                        EnsureDirectoryExistsAndCreateFile().Dispose();
+                        // The results of Metrics and CopyPasteToken analyzers are not needed for Test projects yet the plugin side expects the protobuf files, so we create empty ones.
+                        lock (FileWriteLock)
+                        {
+                            EnsureDirectoryExistsAndCreateFile().Dispose();
+                        }
+
                         return;
                     }
 
                     var messages = c.Compilation.SyntaxTrees
                         .Where(ShouldGenerateMetrics)
                         .Select(x => CreateMessage(x, c.Compilation.GetSemanticModel(x)))
                         .ToArray();
-                    if (messages.Any())
+
+                    lock (FileWriteLock)
                     {
-                        lock (FileWriteLock)
+                        using var metricsStream = EnsureDirectoryExistsAndCreateFile();
+                        foreach (var message in messages)
                         {
-                            using var metricsStream = EnsureDirectoryExistsAndCreateFile();
-                            foreach (var message in messages)
-                            {
-                                message.WriteDelimitedTo(metricsStream);
-                            }
+                            message.WriteDelimitedTo(metricsStream);
                         }
                     }
                 });
 
-        private bool ShouldGenerateMetrics(SyntaxTree tree) =>
+        protected virtual bool ShouldGenerateMetrics(SyntaxTree tree) =>
             FileExtensionWhitelist.Contains(Path.GetExtension(tree.FilePath))
              && (AnalyzeGeneratedCode || !GeneratedCodeRecognizer.IsGenerated(tree));
 
+        protected bool HasTooManyIdentifiers(SyntaxTree syntaxTree, int syntaxKindRaw) =>
+            syntaxTree.GetRoot().DescendantTokens().Count(token => token.RawKind == syntaxKindRaw) > TokenCountThreshold;
```

Comment:
We had an performance issue with `.DescendantNodes()` for large files in `sonar-security` SONARSEC-1603.

If that proves problematic, we can use `CSharpSyntaxWalker` to walk it, count and stop after it's too big.

Reply:
It's related to 8,3 MB file `Models\GlassMapperItemClasses.cs` from the reproducer